packages/agent/src/actor.ts

```diff
@@ -530,25 +530,41 @@ function _createActorMethod(
         arg,
         effectiveCanisterId: ecid,
       });
-
-      requestId;
-      response;
-      requestDetails;
-
-      if (!response.ok || response.body /* IC-1462 */) {
-        throw new UpdateCallRejectedError(cid, methodName, requestId, response);
+      let reply: ArrayBuffer | undefined;
+      let certificate: Certificate | undefined;
+      if (response.body && response.body.certificate) {
+        const cert = response.body.certificate;
+        certificate = await Certificate.create({
+          certificate: bufFromBufLike(cert),
+          rootKey: agent.rootKey as ArrayBuffer,
+          canisterId: Principal.from(canisterId),
+          blsVerify,
+        });
+        const path = [new TextEncoder().encode('request_status'), requestId];
+        const status = new TextDecoder().decode(
+          lookupResultToBuffer(certificate.lookup([...path, 'status'])),
+        );
+
+        switch (status) {
+          case 'replied':
+            reply = lookupResultToBuffer(certificate.lookup([...path, 'reply']));
+            break;
+          case 'rejected':
+            throw new UpdateCallRejectedError(cid, methodName, requestId, response);
+          case 'accepted':
+            // The certificate is not yet ready, so we need to poll for the response
+            break;
+          case 'default':
+            throw new ActorCallError(cid, methodName, 'update', { requestId: toHex(requestId) });
+        }
+      }
+      if (reply === undefined) {
```

Comment:
We should only poll here if the response.status is 202. Otherwise, we can throw the http error received as the request is guaranteed to be not submitted. 

However, the boundary node can be malicious and send whatever they want, so I'm not quite confident on the right behavior here.

Reply:
cc: @robin-kunzler

Reply:
I disagree with this interpretation actually - reply is undefined when it is defined on line 533, and I have set up this handler to handle both the `v2` and `v3` behavior.

`reply` remains undefined if there is no `body` in the response, which is the v2 behavior and is only set if we are able to parse a certificate in the response. 

The Actor is unaware of which strategy the `agent` is using. This is a deliberate choice to maintain compatibility with earlier agents, which are still very popular in use with the system. If a wallet or provider hasn't upgraded, this design will allow the Actor to use an agent that only hits the `v2` endpoint, correctly following up by polling for response.

I can make this clearer in comments, variable names, or so on, however

Reply:
Ah, I missed the v2 fallback but the situation shouldn't change. Even [under v2](https://github.com/dfinity/ic/blob/d436a526d27f528d093f75161a6382c8d2425cc3/rs/http_endpoints/public/src/call/call_v2.rs#L131), `202 ACCEPTED` is the status code from the replica to know that the ingress has been submitted

Reply:
Good to know, I'll make the check more specific in that case!